coderd/fileszip_test.go

```diff
@@ -0,0 +1,217 @@
+			assert.NoError(t, err, "failed to read file %q", path)
+			assert.Equal(t, "hello", string(bs), "unexpected content in file %q", path)
+		case "/test/dir":
+			stat, err := os.Stat(path)
+			assert.NoError(t, err, "failed to stat path %q", path)
+			assert.True(t, stat.IsDir(), "expected path %q to be a directory")
+			if checkModePerm {
+				assert.Equal(t, fs.ModePerm&0o755, stat.Mode().Perm(), "expected mode 0755 on directory")
+			}
+		case "/test/dir/world.txt":
+			stat, err := os.Stat(path)
+			assert.NoError(t, err, "failed to stat path %q", path)
+			assert.False(t, stat.IsDir(), "expected path %q to be a file")
+			if checkModePerm {
+				assert.Equal(t, fs.ModePerm&0o644, stat.Mode().Perm(), "expected mode 0644 on file")
+			}
+			bs, err := os.ReadFile(path)
+			assert.NoError(t, err, "failed to read file %q", path)
+			assert.Equal(t, "world", string(bs), "unexpected content in file %q", path)
+		default:
+			assert.Fail(t, "unexpected path", relPath)
+		}
+
+		t.Logf(relPath)
+		return nil
+	})
+}
+
+func assertSampleTarFile(t *testing.T, tarBytes []byte) {
+	t.Helper()
+
+	tr := tar.NewReader(bytes.NewReader(tarBytes))
+	for {
+		hdr, err := tr.Next()
+		if err != nil {
+			if err == io.EOF {
+				return
+			}
+			require.NoError(t, err)
+		}
+
+		// Note: ignoring timezones here.
+		require.Equal(t, refTime.UTC(), hdr.ModTime.UTC())
+
+		switch hdr.Name {
+		case "test/":
+			require.Equal(t, hdr.Typeflag, byte(tar.TypeDir))
+		case "test/hello.txt":
+			require.Equal(t, hdr.Typeflag, byte(tar.TypeReg))
+			bs, err := io.ReadAll(tr)
+			if err != nil && !xerrors.Is(err, io.EOF) {
+				require.NoError(t, err)
+			}
+			require.Equal(t, "hello", string(bs))
+		case "test/dir/":
+			require.Equal(t, hdr.Typeflag, byte(tar.TypeDir))
+		case "test/dir/world.txt":
+			require.Equal(t, hdr.Typeflag, byte(tar.TypeReg))
+			bs, err := io.ReadAll(tr)
+			if err != nil && !xerrors.Is(err, io.EOF) {
+				require.NoError(t, err)
+			}
+			require.Equal(t, "world", string(bs))
+		default:
+			require.Failf(t, "unexpected file in tar", hdr.Name)
+		}
+	}
+}
+
+func assertSampleZipFile(t *testing.T, zipBytes []byte) {
+	t.Helper()
+
+	zr, err := zip.NewReader(bytes.NewReader(zipBytes), int64(len(zipBytes)))
+	require.NoError(t, err)
+
+	for _, f := range zr.File {
+		// Note: ignoring timezones here.
+		require.Equal(t, refTime.UTC(), f.Modified.UTC())
+		switch f.Name {
+		case "test/", "test/dir/":
+			// directory
+		case "test/hello.txt":
+			rc, err := f.Open()
+			require.NoError(t, err)
+			bs, err := io.ReadAll(rc)
+			_ = rc.Close()
+			require.NoError(t, err)
+			require.Equal(t, "hello", string(bs))
+		case "test/dir/world.txt":
+			rc, err := f.Open()
+			require.NoError(t, err)
+			bs, err := io.ReadAll(rc)
+			_ = rc.Close()
+			require.NoError(t, err)
+			require.Equal(t, "world", string(bs))
+		default:
+			require.Failf(t, "unexpected file in zip", f.Name)
+		}
+	}
+}
+
+func must[T any](t T, err error) T {
```

Comment:
I'd suggest taking T too and using `t.Fatal` instead of panic.

Reply:
moved to a func(*testing.T)